serial_protocol.md

```diff
@@ -0,0 +1,186 @@
+
+##Response Format Byte
+All responses will be prefixed with one of the following bytes signifying the response type.
+```
+  + If response is ok and a raw string value.
+  $ If response is ok, raw string value will be json formatted and prefixed by the byte count and a new line.
+  : If response is ok and a number, this could be float or int.
+  ? If response is ok and a boolean value.
+  - If response is an error
+```
+##NETWORK
+Only needs to be called when the chiplet is in a new environment and needs to connect to a new network. This setting will be stored by the chiplet and persisted through power cycles.
+###Usage
+	NETWORK $SSID
+	NETWORK $SSID $PASSWORD
+###Response
+	CONNECTED - When connected to new network
+	UNABLE_TO_CONNECT - When unable to connect
+###Examples
+	>> NETWORK home-guest
+	<< +CONNECTED
+	>> NETWORK home-private MySecretPassword
+	<< +CONNECTED
+	>> NETWORK home-guest
+	<< -UNABLE_TO_CONNECT
+
+##BEGIN
+Must be called after creating a Serial connection, it can take either just a host for accessing public variables or you may also provide a secret for accessing protected variables in the database.
+###Usage
+	BEGIN $Host
+	BEGIN $Host $Secret
+###Response
+	OK - Accepted initialization parameters
+###Examples
+	>> BEGIN https://samplechat.firebaseio-demo.com
+	<< +OK
+	>> BEGIN https://samplechat.firebaseio-demo.com nnz...sdf
+	<< +OK
+##GET
+Fetches the value at $PATH and returns it on the serial line. If $PATH points to a leaf node you will get the raw value back, if it points to an internal node you will get a JSON string with all children.
+###Usage
+	GET $PATH
+###Response
+	$DATA_AT_PATH
+	$JSON_DATA_BYTE_COUNT \n\r $JSON_DATA
+###Examples
+	>>GET /user/aturing/first
+	<<+Alan
+	>>GET /user/aturing
+	<<$39
+	<<{ "first" : "Alan", "last" : "Turing" }
+	
+##GET{+,*,#,.,?,$}
+Same as GET but will either return the value in the format specified (by the format byte) or return an error.
+###Usage
+	GET+ $PATH
+	GET: $PATH
+	GET? $PATH
+	GET$ $PATH
+###Response
+	$FORMATED_RESPONSE
+###Examples
+	>>GET? /user/aturing/was_human
+	<<?true
+	>>GET? /user/aturing/first
+	<<-ERROR_INCORRECT_FORMAT
+##SET
+Store the data provided at the path provided. This method should be used for simple strings and will assume the first newline is the end of the data.
+###Usage
+	SET $PATH $DATA
+###Response
+	+OK
+	-FAIL
+###Examples
+	>>SET /user/aturing/first Alan
+	<<+OK
+##SET$
+Similar to SET above but used to write multiline strings or raw binary data. Data format is similar to the batch string ($) return type, we specify the $DATA_BYTE_COUNT on the same line as SET$ then a newline and all data. However which the batch string ($) return type returns data json escaped and quoted you may provide raw data and we will handle the escaping.
+
+Receiver will wait until a timeout for client to send $DATA_BYTE_COUNT worth of data before becoming responsive again.
+###Usage
+	SET$ $PATH $DATA_BYTE_COUNT
```

Comment:
I wonder if we should put $DATA_BYTE_COUNT on a separate line, that would make it match `GET$` and also be easier to use from arduino because you could rely on the `Serial.prinln(int)` overload rather than doing string concat or `itoa`.
```
Serial.println("SET$ /foo");
Serial.println(bufSize);
Serial.println(buf);
```